src/tools/tidy/src/no_merge.rs

```diff
@@ -0,0 +1,77 @@
+//! This check makes sure that no accidental merge commits are introduced to the repository.
+//! It forbids all merge commits that are not caused by rollups/bors or subtree syncs.
+
+use std::process::Command;
+
+pub fn check(_: (), bad: &mut bool) {
+    let Ok(remote) = get_rust_lang_rust_remote() else { return; };
+    let Ok(merge_commits) = find_merge_commits(&remote) else { return; };
+
+    let mut bad_merge_commits = merge_commits.lines().filter(|commit| {
+        !(
+            // Bors is the ruler of merge commits.
+            commit.starts_with("Auto merge of")
+                || commit.starts_with("Rollup merge of")
+                // Our fellow subtrees are fine as well.
+                // This can be extended for future subtrees.
+                || commit.contains("Merge from rustc")
+                || commit.contains("merge rustc history")
+                || commit.contains("rust-analyzer")
+                || commit.contains("clippy")
```

Comment:
These concern me. "Merge from rustc" is just a convention, not generated by a tool, and if @RalfJung or @saethlin happen to use a different name for these it will block all Miri syncs unless we hardcode the new string.

Try passing this to git instead:
```
':!src/tools/rust-analyzer' ':!src/tools/clippy' ':!src/tools/rustfmt' ':!src/tools/miri'
```
It might be theoretically possible for someone to accidentally merge those and have it miss the lint, but that seems better than blocking a sync; and I think it's unlikely they'll *only* create a merge commit for the subtree and nothing else.

Actually, we can programatically generate these subtree directories, maybe we should do that instead?
```
; git log | grep git-subtree-dir | awk '{ print $2 }'
src/tools/miri
src/tools/rust-analyzer
compiler/rustc_smir
library/portable-simd
compiler/rustc_codegen_gcc
src/tools/rustfmt
compiler/rustc_codegen_cranelift
src/tools/clippy
```

Reply:
thanks for the offer :) but I don't want to impose that requirement for everyone running syncs, for it to be useful / trustworth all the other tools would need to script it as well, and I think the `:!src/tools/miri` solution is more reliable anyway.

Reply:
Future tools will need to add themselves to this check anyway, we can revisit it then. The `git-subtree-dir` commit is still going to be in the history for miri even if it switches to josh in the future.

Reply:
> ` git log --grep git-subtree-dir | grep git-subtree-dir | awk '{ print $2 }'`

This shell command is pretty non-portable so that's not gonna work and doing the grepping ourselves isn't gonna be amazing either. I think maintaining the list ourselves is fine.

Reply:
Yeah for Miri this works, but future tools might use josh immediately and never use `git subtree`.